doc/group_instances.md

```diff
@@ -0,0 +1,278 @@
+### Group Instance & Group Instance IDs
+
+---
+
+It is recommended that you familiarise yourself with [*repeating routing rules*](repeating_routing_rules.md) before proceeding with this.
+
+Group Instance / Instance IDs are used to identify the instance of an answer acquired from a repeating group.
+
+#### Group Instance (Index)
+An answer can have multiple instances, the *group_instance* tracks this instance and is part of the [*Location*](../app/questionnaire/location.py) object which consists of *group_id*, *group_instance*, and the *block_id*.
+
+Currently, `group_instance` is always used as it is part of the Location. Any non-repeating answers will have a `group_instance` of `0` by default.
+
+##### Group Instance - Repeats
+
+Starting from the first block in the schema, the routing [*path*](../app/questionnaire/path_finder.py) which holds a list of *Location* object is created by following routing rules and evaluating skip conditions until the end of the schema is reached. 
+*The generation of the routing path is carried out on every request.*
+
+When generating the Path, for each group in the schema, we increment the *group_instance* based on the number of times the group repeats. The *no. of repeats* is determined based on the evaluation of the `repeat` rule on each group. 
+
+
+##### Example
+
+```json
+"routing_rules": [{
+    "repeat": {
+        "type": "until",
+        "when": [{
+            "id": "repeating-anyone-else",
+            "condition": "equals",
+            "value": "No"
+        }]
+    }
+}]
+```
+The *no. of repeats* we will have for the above routing rule is based on the number of answers we have with the answer id *repeating-anyone-else* to which the answer is not `"No"`. The *no. of repeats* is incremented for each answer we have with matching id until the routing rule evaluates to *True*.
+If there are no answers yet, it will repeat once.
+
+---
+
+#### Group Instance ID
+
+Unlike `group_instance`, the `group_instance_id` is used only when a *Location* has group dependencies. A `group_instance_id` is a *UUID4* string which represents the instance of an answer. Locations without dependencies are given the default value of `None` for the `group_instance_id`. 
+
+
+##### What defines a group dependency?
+A group is said to have a dependency if it cannot be self sufficient without knowledge of another group/block.
```

Comment:
"self sufficient" is too vague. Something like this would be clearer:

> A group dependency is a mapping of a dependent group name to the block ids that drive that dependency. For example if "group-x" repeats based on an answer in "block-a", then "group-x" is  dependent on the driving "block-a" - `{ 'group-x': ['block-a'] }`.